Implement methods exposing proper sharing completion handling

garnett opened the pull request.

aaustin commented:
Awesome man! We'll review and merge in soon.

danwalkowski commented:
That looks great, thanks!

derrickstaten commented on Branch-SDK/Branch-SDK/BranchUniversalObject.h:
```diff
@@ -39,8 +40,11 @@ typedef NS_ENUM(NSInteger, ContentIndexMode) {
 - (UIActivityItemProvider *)getBranchActivityItemWithLinkProperties:(BranchLinkProperties *)linkProperties;
-- (void)showShareSheetWithShareText:(NSString *)shareText andCallback:(callback)callback;
-- (void)showShareSheetWithLinkProperties:(BranchLinkProperties *)linkProperties andShareText:(NSString *)shareText fromViewController:(UIViewController *)viewController andCallback:(callback)callback;
+- (void)showShareSheetWithShareText:(NSString *)shareText andCallback:(callback)callback __attribute__((deprecated(("This method has been deprecated. Use -[showShareSheetWithShareText:completion:completion] instead."))));
```
the method name in the deprecation notice looks a bit funny here

danwalkowski replied:
indeed

danwalkowski commented on Branch-SDK/Branch-SDK/BranchUniversalObject.m:
```diff
@@ -120,11 +120,11 @@ - (UIActivityItemProvider *)getBranchActivityItemWithLinkProperties:(BranchLinkP
-- (void)showShareSheetWithShareText:(NSString *)shareText andCallback:(callback)callback {
-    [self showShareSheetWithLinkProperties:nil andShareText:shareText fromViewController:nil andCallback:callback];
+- (void)showShareSheetWithShareText:(NSString *)shareText completion:(shareCompletion)completion {
+    [self showShareSheetWithShareText:shareText completion:completion];
```
This line goes into an infinite loop with line 123. Please fix and test.

garnett replied:
Sorry @danwalkowski will fix that soon!

garnett commented:
@danwalkowski I've fixed infinite loop and deprecation message. Also I updated test bed app to use new method (I am happy to write some kind of unit tests but these particular methods are pretty tricky to test properly IMHO)

collinmurray commented:
@garnett Hi there! We're currently using your SDK for branch sharing and have run into the problem that this pull-request solves. Do you guys have an ETA on the deploy time for this?

garnett commented:
Hi @collinmurray, I'm using my fork now, still waiting for maintainers @danwalkowski @aaustin to review this PR

derrickstaten commented on Branch-SDK/Branch-SDK/BranchUniversalObject.m:
```diff
@@ -132,6 +132,16 @@ - (void)showShareSheetWithLinkProperties:(BranchLinkProperties *)linkProperties
     UIActivityViewController *shareViewController = [[UIActivityViewController alloc] initWithActivityItems:items applicationActivities:nil];
     
+    if ([shareViewController respondsToSelector:@selector(completionWithItemsHandler)]) {
+        shareViewController.completionWithItemsHandler = ^(NSString *activityType, BOOL completed, NSArray *returnedItems, NSError *activityError) {
```
Is there a reason why the callback passed (https://github.com/BranchMetrics/iOS-Deferred-Deep-Linking-SDK/pull/270/files#diff-fdcd6f070a770d8a13ad99f764c983e0R14) in doesn't match the callback here? Wouldn't it be cleaner if you could simply have `shareViewController.completionWithItemsHandler = completion`? Just a suggestion

garnett replied:
The reason is that ```UIActivityViewController``` has two callbacks in it's API (iOS6 and iOS8), in order to maintain backward compatibility I've decided to use shorter of those and I'm actually setting it on line 142 as you suggest.

derrickstaten replied:
would this still work on iOS 6 though if you have explicitly listed the newer completion handler format on this line? if the goal is backwards compatibility, is this backward compatible?

garnett replied:
It should. it will work on iOS6, but developer should use iOS SDK 8 or higher in order to compile the code (otherwise Xcode won't recognize `completionWithItemsHandler` SDK method.
New completion handler format consists of simple types so it will work in any case.

danwalkowski commented:
@garnett @collinmurray 
Please consider Derrick's request, and assuming no other issues are found, it will go out in the release next Tuesday.

josh-fuggle commented:
Any update on this?

Sarkar commented:
we just tested this @josh-fuggle - things look good, we'll be cutting a release today or tomorrow with these changes.

Sarkar commented:
FYI @garnett

josh-fuggle commented:
Fantastic, thanks a lot @Sarkar, and thanks for making the PR @garnett
